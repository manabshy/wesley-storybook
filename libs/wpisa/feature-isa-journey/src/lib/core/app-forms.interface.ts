import { CustomerDetailsFormValue } from './customer-details-form-value.interface';
import { ManualAddressFormValue } from '../components/manual-address-form/manual-address-form-value.interface';
import { AddressLookupFormValue } from '../components/address-lookup-form/address-lookup-form-value.interface';
import { LumpSumPaymentFormValue } from '../components/lump-sum-payment-form/lump-sum-payment-form-value.interface';
import { MonthlyPaymentFormValue } from './monthly-payment-form-value.interface';
import { DirectDebitFormValue } from '../components/direct-debit-form/direct-debit-form-value.interface';
import { LumpSumAndMonthlyFormValue } from './lump-sum-and-monthly-form-value.interface';
import { DeclarationFormValue } from './declaration-form-value.interface';

export interface AppForms {
  knowledgeCheckQ1: {
    question1: string;
  };
  knowledgeCheckQ2: {
    question2: string;
  };
  addressLookup: AddressLookupFormValue;
  manualAddress: ManualAddressFormValue;
  customerPersonalDetails: CustomerDetailsFormValue;
  investmentOptions: {
    investmentOption: string;
  };
  lumpSumPayment: LumpSumPaymentFormValue;
  monthlyPayment: MonthlyPaymentFormValue;
  lumpSumAndMonthly: LumpSumAndMonthlyFormValue;
  direcDebit: DirectDebitFormValue;
  declaration: DeclarationFormValue;
}
