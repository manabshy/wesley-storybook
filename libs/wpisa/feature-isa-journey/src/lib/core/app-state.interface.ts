import { CurrentTaxPeriodISALimits } from '@wesleyan-frontend/wpisa/data-access';
import { AppForms } from './app-forms.interface';

export interface AppState {
  attemptId: number;
  currentTaxPeriod: CurrentTaxPeriodISALimits;
  forms: Partial<AppForms>;
}
