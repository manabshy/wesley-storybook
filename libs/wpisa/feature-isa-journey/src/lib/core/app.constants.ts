export const BASE_URL = '/savings-and-investments/with-profits-isa';
