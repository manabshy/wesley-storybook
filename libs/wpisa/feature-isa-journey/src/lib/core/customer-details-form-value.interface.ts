import { AddressLookupFormValue } from '../components/address-lookup-form/address-lookup-form-value.interface';
import { ManualAddressFormValue } from '../components/manual-address-form/manual-address-form-value.interface';

export interface CustomerDetailsFormValue {
  title: string;
  firstName: string;
  lastName: string;
  dob: Dob;
  profession: string;
  nationalInsuranceNumber: string;
  nationality: string;
  addressLookup: AddressLookupFormValue;
  manualAddress: ManualAddressFormValue;
  personalEmail: string;
  personalMobileNumber: string;
  marketingEmail: boolean;
  marketingPost: null;
  marketingPhone: null;
}

export interface Dob {
  day: string;
  month: string;
  year: string;
}
