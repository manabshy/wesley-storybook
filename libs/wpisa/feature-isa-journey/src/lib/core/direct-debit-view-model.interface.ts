export interface DirectDebitViewModel {
  accountNumber: { label: string; value: string };
  sortCode: { label: string; value: string };
}
