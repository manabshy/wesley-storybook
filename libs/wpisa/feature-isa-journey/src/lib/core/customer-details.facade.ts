import { Injectable } from '@angular/core';

import {
  ISAApiService,
  GenericLookupResponse,
  GenericLookup,
  CurrentTaxPeriod,
} from '@wesleyan-frontend/wpisa/data-access';
import { map, take, filter } from 'rxjs/operators';
import { Observable } from 'rxjs';
import { GenericDropdownItem } from './generic-dropdown-item.interface';

@Injectable({
  providedIn: 'root',
})
export class CustomerDetailsFacade {
  genericLookupsResponse$: Observable<GenericLookupResponse>;
  genericLookups$: Observable<GenericLookup[]>;
  currentTaxPeriod$: Observable<CurrentTaxPeriod>;

  titleList$: Observable<GenericDropdownItem[]>;
  nationalityList$: Observable<GenericDropdownItem[]>;
  marketSegmentList$: Observable<GenericDropdownItem[]>;

  constructor(private isaApiService: ISAApiService) {
    this.genericLookupsResponse$ = this.isaApiService.getGenericListAndProductData();
    this.genericLookups$ = this.genericLookupsResponse$.pipe(
      map((data) => data.data.genericLookups)
    );
    this.currentTaxPeriod$ = this.genericLookupsResponse$.pipe(
      map((data) => data.data.currentTaxPeriod)
    );

    this.titleList$ = this.genericLookups$.pipe(
      map((lookups) =>
        lookups.find(({ categoryCode }) => categoryCode === 'TITLE')
      ),
      map((lookup) => this.mapMembersToSelectList(lookup))
    );
    this.nationalityList$ = this.genericLookups$.pipe(
      map((lookups) =>
        lookups.find(({ categoryCode }) => categoryCode === 'NATIONALITY')
      ),
      map((lookup) => this.mapMembersToSelectList(lookup))
    );
    this.marketSegmentList$ = this.genericLookups$.pipe(
      map((lookups) =>
        lookups.find(({ categoryCode }) => categoryCode === 'MARKET_SEGMENT')
      ),
      map((lookup) => this.mapMembersToSelectList(lookup))
    );
  }

  private mapMembersToSelectList(
    lookup: GenericLookup
  ): { value: string; description: string }[] {
    return lookup.members.map((member) => ({
      value: member.memberId,
      description: member.customerDescription,
    }));
  }
}
