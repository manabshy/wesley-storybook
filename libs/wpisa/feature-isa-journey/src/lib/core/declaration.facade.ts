import { Injectable, OnInit } from '@angular/core';
import { NgFormsManager } from '@ngneat/forms-manager';
import {
  Observable,
  of,
  BehaviorSubject,
  combineLatest,
  throwError,
} from 'rxjs';
import {
  map,
  filter,
  tap,
  take,
  startWith,
  catchError,
  concatMap,
  finalize,
} from 'rxjs/operators';
import { format } from 'date-fns';

import {
  ISAApiService,
  ConfigService,
  Declaration,
  CurrentTaxPeriodISALimits,
  SubmitTransactionDTO,
  PaymentType,
  DirectDebitType,
  OnlineDirectDebitDetails,
} from '@wesleyan-frontend/wpisa/data-access';
import { OverlayProgressSpinnerService } from '@wesleyan-frontend/shared/ui-progress-spinner';

import { CustomerDetailsFacade } from './customer-details.facade';
import { AppForms } from './app-forms.interface';
import { CustomerDetailsFormValue } from './customer-details-form-value.interface';
import { PersonalDetailsViewModel } from './personal-details-view-model.interface';
import { ManualAddressFormValue } from '../components/manual-address-form/manual-address-form-value.interface';
import { DirectDebitViewModel } from './direct-debit-view-model.interface';
import { DirectDebitFormValue } from '../components/direct-debit-form/direct-debit-form-value.interface';
import { InvestmentCardViewModel } from './investment-card-view-model.interface';
import { formatCurrencyGBP } from './util-functions';
import {
  InvestmentOptionPaymentTypeStrings,
  InvestmentOptionPaymentType,
} from './investment-option-form-value.interface';
import { isaRoutesNames } from '../isa-journey.routes.names';
import { KnowledgeCheckFacade } from './knowledge-check.facade';

@Injectable({
  providedIn: 'root',
})
export class DeclarationFacade {
  private pageContent: Declaration;
  private pageContentSubject$ = new BehaviorSubject<Declaration>(null);
  private personalDetailsViewModelDataSubject$ = new BehaviorSubject<
    PersonalDetailsViewModel
  >(null);
  private customerPersonalDetailsFormValue: CustomerDetailsFormValue;
  private paymentUrlSubject$ = new BehaviorSubject<string>(null);

  paymentUrl$ = this.paymentUrlSubject$.asObservable();
  pageContent$ = this.pageContentSubject$.asObservable();
  personalDetailsViewModelData$ = this.personalDetailsViewModelDataSubject$.asObservable();
  selectedInvestmentOption$: Observable<InvestmentOptionPaymentTypeStrings>;
  directDebitViewModelData$: Observable<DirectDebitViewModel>;
  investmentCardViewModelData$: Observable<InvestmentCardViewModel>;
  lumpSumAmount$: Observable<number>;
  monthlyAmount$: Observable<number>;

  constructor(
    private isaApiService: ISAApiService,
    private configService: ConfigService,
    private customerDetailsFacade: CustomerDetailsFacade,
    private knowledgeCheckFacade: KnowledgeCheckFacade,
    private formManager: NgFormsManager<AppForms>,
    private loadingService: OverlayProgressSpinnerService
  ) {
    this.pageContent = this.configService.content.declaration;
    this.pageContentSubject$.next(this.pageContent);

    this.customerPersonalDetailsFormValue = this.formManager.getControl(
      'customerPersonalDetails'
    ).value;

    this.personalDetailsViewModelDataSubject$.next(
      this.mapPersonalDetailsFormValuesToViewModel(
        this.customerPersonalDetailsFormValue
      )
    );

    this.selectedInvestmentOption$ = this.formManager.valueChanges(
      'investmentOptions',
      'investmentOption'
    );

    this.directDebitViewModelData$ = this.selectedInvestmentOption$.pipe(
      filter(
        (investmentOption) =>
          investmentOption === InvestmentOptionPaymentType.MONTHLY ||
          investmentOption === InvestmentOptionPaymentType.MONTHLY_AND_LUMP_SUM
      ),
      map((investmentOption) =>
        investmentOption === InvestmentOptionPaymentType.MONTHLY
          ? this.formManager.getControl('monthlyPayment', 'directDebit').value
          : this.formManager.getControl('lumpSumAndMonthly', 'directDebit')
              .value
      ),
      map(this.mapDirectDebitFormValuesToViewModel)
    );

    this.lumpSumAmount$ = this.selectedInvestmentOption$.pipe(
      filter(
        (investmentOption) =>
          investmentOption === InvestmentOptionPaymentType.LUMP_SUM ||
          investmentOption === InvestmentOptionPaymentType.MONTHLY_AND_LUMP_SUM
      ),
      map((investmentOption) =>
        investmentOption === InvestmentOptionPaymentType.LUMP_SUM
          ? this.formManager.getControl('lumpSumPayment', 'amount').value
          : this.formManager.getControl('lumpSumAndMonthly', 'totalAmount')
              .controls.lumpSumAmount.value
      )
    );

    this.monthlyAmount$ = this.selectedInvestmentOption$.pipe(
      filter(
        (investmentOption) =>
          investmentOption === InvestmentOptionPaymentType.MONTHLY ||
          investmentOption === InvestmentOptionPaymentType.MONTHLY_AND_LUMP_SUM
      ),
      map((investmentOption) =>
        investmentOption === InvestmentOptionPaymentType.MONTHLY
          ? this.formManager.getControl('monthlyPayment', 'amount').value
          : this.formManager.getControl('lumpSumAndMonthly', 'totalAmount')
              .controls.monthlyAmount.value
      )
    );

    this.investmentCardViewModelData$ = combineLatest([
      this.lumpSumAmount$.pipe(startWith(0)),
      this.monthlyAmount$.pipe(startWith(0)),
      this.customerDetailsFacade.currentTaxPeriodISALimits$,
    ]).pipe(
      map(([lumpSumAmount, monthlyAmount, isaLimits]) => ({
        showLumpSum: lumpSumAmount ? true : false,
        showMonthly: monthlyAmount ? true : false,
        lumpSumPayment: {
          label: 'Lump-sum amount',
          value: `${formatCurrencyGBP(lumpSumAmount)}`,
        },
        monthlyPayment: {
          label: 'Monthly payments',
          value: `${formatCurrencyGBP(monthlyAmount)}`,
        },
        total: {
          label: this.formatTaxYear(
            this.pageContent.totalInvestmentText,
            isaLimits
          ),
          value: `${formatCurrencyGBP(
            this.calculateTotalInvestment(
              lumpSumAmount,
              monthlyAmount,
              isaLimits.numberOfMonthlyPayments
            )
          )}`,
        },
      }))
    );
  }

  mapDirectDebitFormValuesToViewModel(
    formValue: DirectDebitFormValue
  ): DirectDebitViewModel {
    return {
      accountNumber: {
        label: 'Account number',
        value: formValue.accountNumber,
      },
      sortCode: {
        label: 'Sort code',
        value: `${formValue.sortCode.c1}-${formValue.sortCode.c2}-${formValue.sortCode.c3}`,
      },
    };
  }

  mapPersonalDetailsFormValuesToViewModel(
    formValue: CustomerDetailsFormValue
  ): PersonalDetailsViewModel {
    const {
      title: titleId,
      firstName,
      lastName,
      dob,
      nationality: nationalityId,
      personalEmail,
      profession: professionId,
      personalMobileNumber,
      manualAddress,
    } = formValue;

    const data = {
      name: {
        label: 'Name',
        value: `${titleId} ${firstName} ${lastName}`,
      },
      dob: {
        label: 'Date of birth',
        value: format(
          new Date(
            parseInt(dob.year),
            parseInt(dob.month) - 1, // starts with 0 - January
            parseInt(dob.day)
          ),

          'dd/MM/yyyy'
        ),
      },
      profession: { label: 'Profession', value: professionId },
      nationality: { label: 'Nationality', value: nationalityId },
      emailAddress: { label: 'Email address', value: personalEmail },
      phoneNumber: {
        label: 'Phone number',
        value: personalMobileNumber.replace(/[^0-9]/g, ''),
      },
      homeAddress: {
        label: 'Home address',
        value: this.transformAddress(manualAddress),
      },
    };

    this.customerDetailsFacade
      .getTitleById$(titleId)
      .pipe(
        tap((title) => (data.name.value = `${title} ${firstName} ${lastName}`)),
        take(1)
      )
      .subscribe();

    this.customerDetailsFacade
      .getProfessionById$(professionId)
      .pipe(
        tap((profession) => (data.profession.value = profession)),
        take(1)
      )
      .subscribe();

    this.customerDetailsFacade
      .getNationalityById$(nationalityId)
      .pipe(
        tap((nationality) => (data.nationality.value = nationality)),
        take(1)
      )
      .subscribe();

    return data;
  }

  submitMonthlyISA() {
    this.loadingService.show();

    return this.isaApiService.getTransactionId().pipe(
      map((result) => result.data.transactionId),
      concatMap((transactionId) => {
        const dto = this.getMonthlyTransactionDTO(transactionId);

        return this.isaApiService.submitTransaction(dto).pipe(
          tap((_) => this.loadingService.hide()),
          catchError((err) => {
            window.open(`${isaRoutesNames.KNOWLEDGE_CHECK_ERROR}`, '_self');
            return throwError(err);
          })
        );
      }),
      catchError((err) => {
        window.open(`/${isaRoutesNames.KNOWLEDGE_CHECK_ERROR}`, '_self');
        return throwError(err);
      })
    );
  }

  submitLumpSumISA() {
    this.loadingService.show();

    return this.isaApiService.getTransactionId().pipe(
      map((result) => result.data.transactionId),
      concatMap((transactionId) => {
        const dto = this.getLumpSumTransactionDTO(
          transactionId + Math.random().toString(36).substring(2) //TODO REMOVE RANDOM
        );

        return this.isaApiService.getPaymentUrl(dto).pipe(
          tap(() => this.loadingService.reset()),
          tap((data) => {
            this.paymentUrlSubject$.next(data.paymentUrl);
          }),
          catchError((err) => {
            window.open(`${isaRoutesNames.KNOWLEDGE_CHECK_ERROR}`, '_self');
            return throwError(err);
          })
        );
      }),
      catchError((err) => {
        window.open(`/${isaRoutesNames.KNOWLEDGE_CHECK_ERROR}`, '_self');
        return throwError(err);
      }),
      finalize(() => this.loadingService.hide())
    );
  }

  submitLumpSumAndMonthlyISA() {
    this.loadingService.show();

    return this.isaApiService.getTransactionId().pipe(
      map((result) => result.data.transactionId),
      concatMap((transactionId) => {
        const dto = this.getLumpSumAndMonthlyTransactionDTO(
          transactionId + Math.random().toString(36).substring(2) //TODO REMOVE RANDOM
        );

        return this.isaApiService.getPaymentUrl(dto).pipe(
          tap(() => this.loadingService.reset()),
          tap((data) => {
            this.paymentUrlSubject$.next(data.paymentUrl);
          }),
          tap((_) => this.loadingService.hide()),
          catchError((err) => {
            window.open(`${isaRoutesNames.KNOWLEDGE_CHECK_ERROR}`, '_self');
            return throwError(err);
          })
        );
      }),
      catchError((err) => {
        console.log('isa both', err);
        window.open(`/${isaRoutesNames.KNOWLEDGE_CHECK_ERROR}`, '_self');
        return throwError(err);
      }),
      finalize(() => this.loadingService.hide())
    );
  }

  getMonthlyTransactionDTO(
    transactionId: string
  ): Omit<SubmitTransactionDTO, 'lumpPaymentDetails' | 'lumpAmount'> {
    const mappedCustomerDTO = this.customerDetailsFacade.mapCustomerFormToSearchCustomerDTO(
      this.formManager.getControl('customerPersonalDetails').value
    );

    const dto = {
      currentAddress: mappedCustomerDTO.currentAddress,
      customerDetails: mappedCustomerDTO.customerDetails,
      nationalityDetails: mappedCustomerDTO.nationalityDetails,
      marketingPreferences: mappedCustomerDTO.marketingPreferences,
      transactionId: transactionId,
      appTestAttemptId: this.knowledgeCheckFacade.knowledgeCheckAttemptId,
      taxPeriodCode: '',
      paymentType: PaymentType.Regular,
      regularAmount: this.formManager.getControl('monthlyPayment', 'amount')
        .value,
      directDebitType: DirectDebitType.ONLINE,
      onlineDirectDebitDetails: this.mapDirectDebit(
        this.formManager.getControl('monthlyPayment', 'directDebit').value
      ),
      customerPermissionGranted: true,
    };

    this.customerDetailsFacade.currentTaxPeriodISALimits$
      .pipe(
        tap((value) => (dto.taxPeriodCode = value.taxPeriodCode)),
        take(1)
      )
      .subscribe();
    return dto;
  }

  getLumpSumTransactionDTO(
    transactionId: string
  ): Omit<
    SubmitTransactionDTO,
    'directDebitType' | 'regularAmount' | 'onlineDirectDebitDetails'
  > {
    const mappedCustomerDTO = this.customerDetailsFacade.mapCustomerFormToSearchCustomerDTO(
      this.formManager.getControl('customerPersonalDetails').value
    );

    const dto = {
      currentAddress: mappedCustomerDTO.currentAddress,
      customerDetails: mappedCustomerDTO.customerDetails,
      nationalityDetails: mappedCustomerDTO.nationalityDetails,
      marketingPreferences: mappedCustomerDTO.marketingPreferences,
      transactionId: transactionId,
      appTestAttemptId: this.knowledgeCheckFacade.knowledgeCheckAttemptId,
      taxPeriodCode: '',
      paymentType: PaymentType.LumpSum,
      lumpAmount: this.formManager.getControl('lumpSumPayment', 'amount').value,
      customerPermissionGranted: true,
    };

    this.customerDetailsFacade.currentTaxPeriodISALimits$
      .pipe(
        tap((value) => (dto.taxPeriodCode = value.taxPeriodCode)),
        take(1)
      )
      .subscribe();
    return dto;
  }

  getLumpSumAndMonthlyTransactionDTO(
    transactionId: string
  ): SubmitTransactionDTO {
    const mappedCustomerDTO = this.customerDetailsFacade.mapCustomerFormToSearchCustomerDTO(
      this.formManager.getControl('customerPersonalDetails').value
    );

    const dto = {
      currentAddress: mappedCustomerDTO.currentAddress,
      customerDetails: mappedCustomerDTO.customerDetails,
      nationalityDetails: mappedCustomerDTO.nationalityDetails,
      marketingPreferences: mappedCustomerDTO.marketingPreferences,
      transactionId: transactionId,
      appTestAttemptId: this.knowledgeCheckFacade.knowledgeCheckAttemptId,
      taxPeriodCode: '',
      paymentType: PaymentType.Both,
      lumpAmount: this.formManager.getControl(
        'lumpSumAndMonthly',
        'totalAmount.lumpSumAmount'
      ).value,
      regularAmount: this.formManager.getControl(
        'lumpSumAndMonthly',
        'totalAmount.monthlyAmount'
      ).value,
      directDebitType: DirectDebitType.ONLINE,
      onlineDirectDebitDetails: this.mapDirectDebit(
        this.formManager.getControl('lumpSumAndMonthly', 'directDebit').value
      ),
      customerPermissionGranted: true,
    };

    this.customerDetailsFacade.currentTaxPeriodISALimits$
      .pipe(
        tap((value) => (dto.taxPeriodCode = value.taxPeriodCode)),
        take(1)
      )
      .subscribe();
    return dto;
  }

  mapDirectDebit(value: DirectDebitFormValue): OnlineDirectDebitDetails {
    return {
      accountHolderFullName: value.accountHolderFullName,
      sortCode: `${value.sortCode.c1}-${value.sortCode.c2}-${value.sortCode.c3}`,
      accountNumber: value.accountNumber,
      bankName: value.bankName,
      customerConfimedDD: true,
      accountValidated: false,
    };
  }

  calculateTotalInvestment(
    lumpSumAmount: number = 0,
    monthlyAmount: number,
    numberOfMonthlyPayments: number
  ): number {
    return (
      parseFloat('' + lumpSumAmount) +
      parseFloat('' + monthlyAmount) * parseInt('' + numberOfMonthlyPayments)
    );
  }

  transformAddress(address: ManualAddressFormValue, delimiter = ', '): string {
    if (!address) {
      return '';
    }

    const fields: { fieldName: string; delimited: boolean }[] = [
      { fieldName: 'flatNumber', delimited: false },
      { fieldName: 'houseNumber', delimited: false },
      { fieldName: 'houseName', delimited: true },
      { fieldName: 'street', delimited: true },
      { fieldName: 'town', delimited: true },
      { fieldName: 'region', delimited: true },
      { fieldName: 'county', delimited: true },
      { fieldName: 'postcode', delimited: true },
    ];

    let result = fields.reduce((accumulator, field) => {
      const value = address[field.fieldName];

      return value
        ? `${accumulator}${value}${field.delimited ? delimiter : ' '}`
        : accumulator;
    }, '');

    result = result.substr(0, result.length - delimiter.length);

    return result;
  }

  formatTaxYear(text: string, isaLimits: CurrentTaxPeriodISALimits): string {
    return text.replace(
      '{tax-year}',
      format(new Date(isaLimits.startDateTime), 'yyyy') +
        '/' +
        format(new Date(isaLimits.endDateTime), 'yy')
    );
  }
}
