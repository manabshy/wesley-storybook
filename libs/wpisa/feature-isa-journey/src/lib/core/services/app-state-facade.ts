import { Injectable } from '@angular/core';
import { mergeDeepRight, reduce, omit, find, propEq } from 'ramda';
import { NgFormsManager } from '@ngneat/forms-manager';
import { map } from 'rxjs/operators';
import { Observable, of } from 'rxjs';

import { SessionStorageService } from '@wesleyan-frontend/wpisa/data-access';

import { AppForms } from '../models/app-forms.interface';
import { AppState } from '../models/app-state.interface';

@Injectable({
  providedIn: 'root',
})
export class AppStateFacade {
  private appState: AppState;
  state$ = this.sessionStorageService.appState$.pipe(
    map((state) => JSON.parse(state))
  );

  constructor(
    private sessionStorageService: SessionStorageService,
    private formManager: NgFormsManager<AppForms>
  ) {
    this.state$.subscribe((s) => (this.appState = s));
  }

  save(state: Partial<AppState>): Observable<{}> {
    this.appState = mergeDeepRight(this.appState, state);

    return this.sessionStorageService.save(JSON.stringify(this.appState));
  }

  saveFormState(
    formName: keyof AppForms | Array<keyof AppForms>,
    omitValues?: { formName: keyof AppForms; controlName: string[] }[]
  ) {
    const listOfControls = Array.isArray(formName) ? formName : [formName];

    const formStateObj: Partial<AppForms> = this.processObject(
      omitValues,
      listOfControls
    );

    return this.save({
      forms: {
        ...formStateObj,
      },
    });
  }

  private processObject(
    omitValues: { formName: keyof AppForms; controlName: string[] }[],
    listOfControls: Array<keyof AppForms>
  ): Partial<AppForms> {
    return reduce(
      (acc: Partial<AppForms>, formKey: keyof AppForms) => {
        const formValue = this.formManager.getControl(formKey)?.value;
        const valuesToOmit = omitValues
          ? find(propEq('formName', formKey))(omitValues)
          : undefined;

        return {
          ...acc,
          [formKey]: valuesToOmit
            ? omit(valuesToOmit.controlName, formValue)
            : formValue,
        };
      },
      {},
      listOfControls
    );
  }

  get state() {
    return this.appState;
  }
}
