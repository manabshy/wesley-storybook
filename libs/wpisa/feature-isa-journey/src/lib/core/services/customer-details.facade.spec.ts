import { TestBed } from '@angular/core/testing';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { Overlay } from '@angular/cdk/overlay';
import { of } from 'rxjs';

import {
  SearchCustomerDTO,
  ISAApiService,
  ConfigService,
} from '@wesleyan-frontend/wpisa/data-access';
import { OverlayProgressSpinnerService } from '@wesleyan-frontend/shared/ui-progress-spinner';
import {
  mockConfig,
  mockGenericLookups,
} from '@wesleyan-frontend/wpisa/data-access';

import { CustomerDetailsFacade } from './customer-details.facade';
import { AppStateFacade } from './app-state-facade';

describe('CustomerDetailsFacade', () => {
  let customerDetailsFacade: CustomerDetailsFacade;
  const mockCustomerDetailsFormSubmit = {
    title: mockConfig.titleList[1].value,
    firstName: 'Alex',
    lastName: 'Cote',
    dob: {
      day: '01',
      month: '02',
      year: '1980',
    },
    profession: 'DENTISTS',
    nationalInsuranceNumber: 'SN 234 234 A',
    nationality: '119',
    addressLookup: {
      postcode: null,
      selectedAddressId: null,
    },
    manualAddress: {
      flatNumber: null,
      houseNumber: '9',
      houseName: null,
      street: 'Middle Street',
      town: 'Rowley',
      region: 'West Midlands',
      county: null,
      postcode: 'B62 9HY',
    },
    personalEmail: 'alex.cotelin@wesleyan.co.uk',
    personalMobileNumber: '07553 770 986',
    marketingEmail: true,
    marketingPost: null,
    marketingPhone: null,
  };

  const expected: SearchCustomerDTO = {
    customerDetails: {
      title: 'Mrs',
      forename: 'Alex',
      surname: 'Cote',
      dateOfBirth: '1980-02-01',
      mobilePhoneNumber: '07553770986',
      niNumber: 'SN234234A',
      emailAddress: 'alex.cotelin@wesleyan.co.uk',
      marketSegmentCode: 'DENTISTS',
    },
    currentAddress: {
      flat: null,
      houseName: null,
      houseNumber: '9',
      street: 'Middle Street',
      town: 'Rowley',
      district: 'West Midlands',
      county: null,
      postcode: 'B62 9HY',
    },
    nationalityDetails: {
      primaryNationality: 'French',
    },
    marketingPreferences: {
      permitForPost: null,
      permitForTelephone: null,
      permitForEmail: true,
    },
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule],
      providers: [
        AppStateFacade,
        OverlayProgressSpinnerService,
        Overlay,
        {
          provide: ConfigService,
          useValue: { content: mockConfig },
        },
        {
          provide: ISAApiService,
          useValue: {
            getGenericListAndProductData: () => of(mockGenericLookups),
          },
        },
      ],
    });

    customerDetailsFacade = TestBed.inject(CustomerDetailsFacade);
  });

  it('should be created', () => {
    expect(customerDetailsFacade).toBeTruthy();
  });

  describe('mapCustomerFormToSearchCustomerDTO', () => {
    it('should map customer form values to SearchCustomerDTO', () => {
      const result = customerDetailsFacade.mapCustomerFormToSearchCustomerDTO(
        mockCustomerDetailsFormSubmit
      );

      expect(result).toEqual(expected);
    });
  });

  //   describe('mapCustomerDetails', () => {
  //     it('should map customer form values to customer details search DTO', () => {
  //       const result = service.mapCustomerDetails(mockCustomerDetailsFormSubmit);

  //       expect(result).toEqual(expected.customerDetails);
  //     });
  //   });

  //   describe('mapCustomerAddress', () => {
  //     it('should map customer form values to CurrentAddressDTO', () => {
  //       const result = service.mapCustomerAddress(mockCustomerDetailsFormSubmit);

  //       expect(result).toEqual(expected.currentAddress);
  //     });
  //   });

  //   describe('mapCustomerMarketingPreferences', () => {
  //     it('should map customer form values to MarketingPreferencesDTO', () => {
  //       const result = service.mapCustomerMarketingPreferences(
  //         mockCustomerDetailsFormSubmit
  //       );

  //       expect(result).toEqual(expected.marketingPreferences);
  //     });
  //   });

  //   describe('mapCustomerNationality', () => {
  //     it('should map customer form values to NationalityDTO', () => {
  //       const result = service.mapCustomerNationality(
  //         mockCustomerDetailsFormSubmit
  //       );

  //       expect(result).toEqual(expected.nationalityDetails);
  //     });
  //   });
});
