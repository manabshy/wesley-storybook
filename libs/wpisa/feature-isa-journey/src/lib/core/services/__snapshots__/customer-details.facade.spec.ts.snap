// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetailsFacade marketSegmentList$ should map marketSegmentList 1`] = `
Array [
  Object {
    "description": "GP",
    "value": "GENMEDPRAC",
  },
  Object {
    "description": "Hospital Doctor",
    "value": "HOSDOCS",
  },
  Object {
    "description": "Student Medic",
    "value": "TBD",
  },
  Object {
    "description": "Dentist",
    "value": "DENTISTS",
  },
  Object {
    "description": "Student Dentist",
    "value": "DENTALSTU",
  },
  Object {
    "description": "Teacher",
    "value": "TEACHERS",
  },
  Object {
    "description": "Lawyer",
    "value": "WFL",
  },
  Object {
    "description": "Other",
    "value": "OTHER",
  },
]
`;

exports[`CustomerDetailsFacade nationalityList$ should map nationality 1`] = `
Array [
  Object {
    "description": "British",
    "value": "92",
  },
  Object {
    "description": "American",
    "value": "95",
  },
  Object {
    "description": "Australian",
    "value": "159",
  },
  Object {
    "description": "Dutch",
    "value": "113",
  },
  Object {
    "description": "French",
    "value": "119",
  },
  Object {
    "description": "German",
    "value": "121",
  },
]
`;

exports[`CustomerDetailsFacade titleList$ should map titles 1`] = `
Array [
  Object {
    "description": "Mr",
    "value": "c2044dac-a1eb-48c8-b23c-5255bac96bc0",
  },
  Object {
    "description": "Mrs",
    "value": "37c037af-7ecb-47a9-8c16-738150d57d26",
  },
  Object {
    "description": "Miss",
    "value": "b0fae53c-fc94-4a82-9b92-426e4b54485f",
  },
  Object {
    "description": "Ms",
    "value": "d7743753-25ff-4343-820d-c1a45ce39d25",
  },
  Object {
    "description": "Dr",
    "value": "ad28900f-c863-4ea3-bbcf-83aca8065533",
  },
  Object {
    "description": "Professor",
    "value": "ae9cde20-ab54-4b07-ba80-933b487729f6",
  },
]
`;
